gdal/frmts/netcdf/netcdfdataset.cpp

```diff
@@ -10556,3 +10111,824 @@ static char **NCDFTokenizeArray( const char *pszValue )
+    NCDF_ERR_RET(nc_inq_grps(nGroupId, nullptr, panSubGroupIds));
+    *pnSubGroups = nSubGroups;
+    *ppanSubGroupIds = panSubGroupIds;
+#endif
+
+    return CE_None;
+}
+
+// Get the full name of a given NetCDF (or group) ID
+// (e.g. /group1/group2/.../groupn).
+// bNC3Compat remove the leading slash for top-level variables for
+// backward compatibility (top-level variables are the ones in the root group).
+static CPLErr NCDFGetGroupFullName( int nGroupId, char **ppszFullName,
+                                    bool bNC3Compat )
+{
+    *ppszFullName = nullptr;
+
+#ifdef NETCDF_HAS_NC4
+    size_t nFullNameLen;
+    NCDF_ERR_RET(nc_inq_grpname_len(nGroupId, &nFullNameLen));
+    *ppszFullName = (char *) CPLMalloc((nFullNameLen + 1) * sizeof(char));
+    int status = nc_inq_grpname_full(nGroupId, &nFullNameLen, *ppszFullName);
+    if( status != NC_NOERR )
+    {
+        CPLFree(*ppszFullName);
+        *ppszFullName = nullptr;
+        NCDF_ERR_RET(status);
+    }
+#else
+    *ppszFullName = CPLStrdup("/");
+#endif
+
+    if( bNC3Compat && EQUAL(*ppszFullName, "/") )
+        (*ppszFullName)[0] = '\0';
+
+    return CE_None;
+}
+
+// Get the full name of a given NetCDF variable ID
+// (e.g. /group1/group2/.../groupn/var).
+// Handle also NC_GLOBAL as nVarId.
+// bNC3Compat remove the leading slash for top-level variables for
+// backward compatibility (top-level variables are the ones in the root group).
+static CPLErr NCDFGetVarFullName( int nGroupId, int nVarId,
+                                  char **ppszFullName, bool bNC3Compat )
+{
+    *ppszFullName = nullptr;
+    char *pszGroupFullName = nullptr;
+    ERR_RET(NCDFGetGroupFullName(nGroupId, &pszGroupFullName, bNC3Compat));
+    char szVarName[NC_MAX_NAME + 1];
+    if( nVarId == NC_GLOBAL )
+    {
+        strcpy(szVarName, "NC_GLOBAL");
+    }
+    else
+    {
+        int status = nc_inq_varname(nGroupId, nVarId, szVarName);
+        if( status != NC_NOERR )
+        {
+            CPLFree(pszGroupFullName);
+            NCDF_ERR_RET(status);
+        }
+    }
+    const char *pszSep = "/";
+    if( EQUAL(pszGroupFullName, "/") || EQUAL(pszGroupFullName, "") )
+        pszSep = "";
+    *ppszFullName = CPLStrdup(CPLSPrintf("%s%s%s", pszGroupFullName, pszSep,
+                                         szVarName));
+    CPLFree(pszGroupFullName);
+    return CE_None;
+}
+
+// Get the NetCDF root group ID of a given group ID.
+static CPLErr NCDFGetRootGroup( int nStartGroupId, int *pnRootGroupId )
+{
+    *pnRootGroupId = -1;
+#ifdef NETCDF_HAS_NC4
+    // Recurse on parent group.
+    int nParentGroupId;
+    int status = nc_inq_grp_parent(nStartGroupId, &nParentGroupId);
+    if( status == NC_NOERR )
+        return NCDFGetRootGroup(nParentGroupId, pnRootGroupId);
+    else if( status != NC_ENOGRP )
+        NCDF_ERR_RET(status);
+    else // No more parent group.
+#endif
+        *pnRootGroupId = nStartGroupId;
```

Comment:
would be more robust and less odd regarding indentation to surround this line by { and }